{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Change intermediate isoform counts per allele to ESPRESSO format\n",
    "## so we can use it as input to the DRIMSeq DTU script.\n",
    "## format: transcript_ID, transcript_name, gene_ID, sample_name1, sample_name2, ...\n",
    "\n",
    "def interm_counts_to_espresso_format(file_path,\n",
    "                                     gene_ID,\n",
    "                                     transcript_ID,\n",
    "                                     transcript_name,\n",
    "                                     columns,\n",
    "                                     values):\n",
    "    df = pd.read_table(file_path)\n",
    "\n",
    "    df['gene_ID'] = df[gene_ID].astype(str).agg('/'.join, axis=1)\n",
    "    df['transcript_ID'] = df[transcript_ID].astype(str).agg('/'.join, axis=1)\n",
    "    df['transcript_name'] = df[transcript_name].astype(str).agg('/'.join, axis=1) # same as transcript_ID\n",
    "\n",
    "    esp_df = df.pivot_table(index=['transcript_ID', 'transcript_name', 'gene_ID'],\n",
    "                            columns=columns,\n",
    "                            values=values,\n",
    "                            fill_value=0)\n",
    "    \n",
    "    esp_df.columns = esp_df.columns.values\n",
    "    esp_df.reset_index(inplace=True)\n",
    "\n",
    "    return esp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Format counts to espresso format and keep replicates separate\n",
    "\n",
    "indir = '../data/splicing_order_stringent_version/replicates_separate'\n",
    "files = [os.path.relpath(os.path.join(indir, f)) for f in os.listdir(indir)]\n",
    "\n",
    "outdir = '../data/splicing_order_stringent_version/replicates_separate_esp_format'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "# Get all cell lines\n",
    "cell_lines = set()\n",
    "for file in files:\n",
    "    cell_line = re.findall(r'(GM\\d{5})', file)\n",
    "    cell_lines.update(cell_line)\n",
    "\n",
    "# Associate each file to its cell line\n",
    "cell_line_files_dict = {cell_line: [] for cell_line in cell_lines}\n",
    "for file in files:\n",
    "    for cell_line in cell_lines:\n",
    "        if cell_line in file:\n",
    "            cell_line_files_dict[cell_line].append(file)\n",
    "\n",
    "# Format each file like espresso output\n",
    "for cell_line in cell_line_files_dict:\n",
    "    dfs = []\n",
    "    for file in cell_line_files_dict[cell_line]:\n",
    "        # Yield three possible intermediate isoforms per splcing level\n",
    "        df = interm_counts_to_espresso_format(\n",
    "            file_path=file,\n",
    "            gene_ID=['gene', 'analyzed_introns', 'level'],\n",
    "            transcript_ID=['gene_ID', 'pattern2'],\n",
    "            transcript_name=['gene_ID', 'pattern2'],\n",
    "            columns='allele',\n",
    "            values='count_pattern2'\n",
    "        )\n",
    "        dfs.append(df)\n",
    "\n",
    "    # Concatenate each replicates of a cell line into one file\n",
    "    combined_df = pd.DataFrame()\n",
    "    for i, df in enumerate(dfs):\n",
    "        df = df.rename(columns={'M': f'M_{i+1}', 'P': f'P_{i+1}'})\n",
    "\n",
    "        if i == 0:\n",
    "            combined_df = df\n",
    "        else:\n",
    "            combined_df = pd.merge(combined_df, df, on=['transcript_ID', 'transcript_name', 'gene_ID'], how='outer')\n",
    "        \n",
    "    combined_df = combined_df.fillna(0)\n",
    "    fname = os.path.join(outdir, f'{cell_line}_interm_counts_per_allele.hac.all_introns.min10reads.filterND.stringent.3_isoforms.esp.txt')\n",
    "    combined_df.to_csv(fname, index=None, sep='\\t')\n",
    "\n",
    "    # Necessary group files for DRIMSeq script\n",
    "    num_reps = len(cell_line_files_dict[cell_line])\n",
    "    group_1 = ','.join([f'M_{i}' for i in range(1, num_reps+1)]) + '\\n'\n",
    "    group_2 = ','.join([f'P_{i}' for i in range(1, num_reps+1)]) + '\\n'\n",
    "    \n",
    "    fname = os.path.join(outdir, f'{cell_line}_interm_counts_per_allele.hac.all_introns.min10reads.filterND.stringent.3_isoforms.group_1.txt')\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write(group_1)\n",
    "\n",
    "    fname = os.path.join(outdir, f'{cell_line}_interm_counts_per_allele.hac.all_introns.min10reads.filterND.stringent.3_isoforms.group_2.txt')\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write(group_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Format counts to espresso format and merge replicates\n",
    "\n",
    "indir = '../data/splicing_order_stringent_version/merged_replicates'\n",
    "files = [os.path.relpath(os.path.join(indir, f)) for f in os.listdir(indir)]\n",
    "\n",
    "outdir = '../data/splicing_order_stringent_version/merged_replicates_esp_format'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "for file in files:\n",
    "    df = interm_counts_to_espresso_format(\n",
    "        # Yield three possible intermediate isoforms per splcing level\n",
    "        file_path=file,\n",
    "            gene_ID=['gene', 'analyzed_introns', 'level'],\n",
    "            transcript_ID=['gene_ID', 'pattern2'],\n",
    "            transcript_name=['gene_ID', 'pattern2'],\n",
    "            columns='allele',\n",
    "            values='count_pattern2'\n",
    "    )\n",
    "                      \n",
    "    directory, file_name = os.path.split(file)\n",
    "    base_name, ext = os.path.splitext(file_name)\n",
    "    df_fname = os.path.join(outdir, f\"{base_name}.3_isoforms.esp.txt\")                                        \n",
    "\n",
    "    df.to_csv(df_fname, index=None, sep='\\t')\n",
    "\n",
    "    group1_fname = os.path.join(outdir, f\"{base_name}.3_isoforms.group_1.txt\") \n",
    "    group2_fname = os.path.join(outdir, f\"{base_name}.3_isoforms.group_2.txt\") \n",
    "\n",
    "    with open(group1_fname, 'w') as f:\n",
    "        f.write('M')\n",
    "    with open(group2_fname, 'w') as f:\n",
    "        f.write('P')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
