{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format espresso counts\n",
    "indir = '../data/splicing_order_stringent_version/merged_replicates_esp_format'\n",
    "files = [os.path.relpath(os.path.join(indir, f)) for f in os.listdir(indir)]\n",
    "files = [f for f in files if 'esp.txt' in f]\n",
    "\n",
    "outdir = '../data/drimseq_tuqtl_input'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "dfs = []\n",
    "for file in files:\n",
    "    cell_line = re.findall(r'(GM\\d{5})', file)[0]\n",
    "    df = pd.read_table(file)\n",
    "    df = df[['transcript_ID', 'gene_ID', 'M', 'P']]\n",
    "    df.columns = ['feature_id', 'gene_id', f'{cell_line + '_M'}', f'{cell_line + '_P'}']\n",
    "    dfs.append(df)\n",
    "\n",
    "merged_df = dfs[0]\n",
    "for df in dfs[1:]:\n",
    "    merged_df = pd.merge(merged_df, df, on=['feature_id', 'gene_id'], how='outer')\n",
    "merged_df = merged_df.fillna(0)\n",
    "merged_df.to_csv(os.path.join(outdir, 'LCL_interm_counts_tuqtl.tsv'),\n",
    "                 index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format genotypes\n",
    "genotypes_path = '../data/drimseq_tuqtl_input/LCLs_studied_cell_lines.genes_splicing_order.extended.formatted_for_drimseq.txt'\n",
    "genotypes_df = pd.read_table(genotypes_path)\n",
    "genotypes_df = genotypes_df.drop(columns=['REF', 'ALT'])\n",
    "genotypes_df = genotypes_df.rename(columns={\n",
    "    'CHROM': 'chr', 'START': 'start', 'ID': 'snp_id'}\n",
    ")\n",
    "genotypes_df.insert(loc=2, column='end', value=genotypes_df['start'])\n",
    "genotypes_df.to_csv('../data/drimseq_tuqtl_input/LCL_genotypes.tsv', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix gene ids of the bed file so that they are the same as the unique ids of the counts file\n",
    "bed = pd.read_table('../data/drimseq_tuqtl_input/NCBI_RefSeq_hg38_genes_parsed.bed', header=None)\n",
    "merged_interm_counts = pd.read_table('../data/drimseq_tuqtl_input/LCL_interm_counts_tuqtl.tsv')\n",
    "\n",
    "# Special unique gene ids\n",
    "uniq_gene_id = pd.Series(merged_interm_counts['gene_id'].unique())\n",
    "gene_id = uniq_gene_id.str.split('/').str[0]\n",
    "# Keep only wanted genes\n",
    "bed = bed[bed.iloc[:,3].isin(gene_id)]\n",
    "\n",
    "# Standard gene id and unique gene id in a dictionary\n",
    "gene_id_dict = defaultdict(list)\n",
    "\n",
    "for key, value in zip(gene_id, uniq_gene_id):\n",
    "    gene_id_dict[key].append(value)\n",
    "\n",
    "# Make new bed from unique gene id\n",
    "new_rows_list = []\n",
    "for i, old_gene_id in enumerate(bed.iloc[:,3]):\n",
    "    if old_gene_id in gene_id_dict:\n",
    "        for new_gene_id in gene_id_dict[old_gene_id]:\n",
    "            new_row = bed.iloc[i].copy()\n",
    "            new_row[3] = new_gene_id\n",
    "            new_rows_list.append(new_row)\n",
    "\n",
    "new_df = pd.DataFrame(new_rows_list)\n",
    "\n",
    "new_df.to_csv('../data/drimseq_tuqtl_input/NCBI_RefSeq_hg38_genes_parsed_uniq_id.bed',\n",
    "              index=None, sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix counts\n",
    "def interm_counts_to_espresso_format(file_path,\n",
    "                                     gene_ID,\n",
    "                                     transcript_ID,\n",
    "                                     transcript_name,\n",
    "                                     columns,\n",
    "                                     values):\n",
    "    df = pd.read_table(file_path)\n",
    "\n",
    "    df['gene_ID'] = df[gene_ID].astype(str).agg('/'.join, axis=1)\n",
    "    df['transcript_ID'] = df[transcript_ID].astype(str).agg('/'.join, axis=1)\n",
    "    df['transcript_name'] = df[transcript_name].astype(str).agg('/'.join, axis=1) # same as transcript_ID\n",
    "\n",
    "    esp_df = df.pivot_table(index=['transcript_ID', 'transcript_name', 'gene_ID'],\n",
    "                            columns=columns,\n",
    "                            values=values,\n",
    "                            fill_value=0)\n",
    "    \n",
    "    esp_df.columns = esp_df.columns.values\n",
    "    esp_df.reset_index(inplace=True)\n",
    "\n",
    "    return esp_df\n",
    "\n",
    "\n",
    "\n",
    "df = interm_counts_to_espresso_format(\n",
    "    file_path='../data/splicing_order_stringent_version/LCLs_merged_interm_counts_per_allele_filtered_min_2_sig_cell_lines.txt',\n",
    "    gene_ID=['gene', 'analyzed_introns', 'level'],\n",
    "    transcript_ID=['gene', 'pattern2'],\n",
    "    transcript_name=['gene', 'pattern2'],\n",
    "    columns=['cell_line', 'allele'],\n",
    "    values='count_pattern2'\n",
    ")\n",
    "\n",
    "df.columns = ['_'.join(col) if isinstance(col, tuple) else col for col in df.columns]\n",
    "df = df.drop(columns='transcript_name')\n",
    "df = df.rename(columns={\n",
    "    'transcript_ID': 'feature_id',\n",
    "    'gene_ID': 'gene_id'\n",
    "}\n",
    ")\n",
    "\n",
    "df.to_csv('../data/drimseq_tuqtl_input/LCLs_merged_interm_counts_per_allele_filtered_min_2_sig_cell_lines_3_isoforms.tsv',\n",
    "          sep='\\t', index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
