{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f99269",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "This script will analyze the output from transcript quantification with espresso and identify AS events\n",
    "that overlap with genes with allele-specific splicing orders\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a0ce790d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import itertools\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5266365f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get espresso results\n",
    "esp_df = pd.read_table(\"/path/to/results_files/LCL_espresso_counts.tsv\")\n",
    "\n",
    "# Get gene_ID to gene_name correspondence\n",
    "ann_df = pd.read_table(\"/path/to/Ensembl_geneID2name_hsapiens_gene_ensembl.txt\")\n",
    "\n",
    "# Merge espresso results with gene_ID/gene_name\n",
    "esp_df = esp_df.merge(ann_df[['ensembl_gene_id','external_gene_name']].drop_duplicates(), left_on=['gene_ID'], right_on=['ensembl_gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1e4ca54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat espresso results\n",
    "esp_df_m = esp_df.melt(id_vars=['transcript_ID','transcript_name','gene_ID','ensembl_gene_id','external_gene_name'],\n",
    "           var_name = 'allele_tmp', value_name = 'count')\n",
    "esp_df_m['cell_line'] = esp_df_m['allele_tmp'].str.split(\"_\").str[0]\n",
    "esp_df_m['allele'] = esp_df_m['allele_tmp'].str.split(\"\\\\.\").str[1]\n",
    "esp_df_grp = pd.DataFrame(esp_df_m.groupby(['transcript_ID','gene_ID','external_gene_name','cell_line','allele'])['count'].sum()).reset_index()\n",
    "esp_df_piv = esp_df_m.pivot_table(index=['transcript_ID','gene_ID','external_gene_name','cell_line'],columns='allele',values='count').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb9e7a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get genes for which we computed splicing order\n",
    "order_df = pd.read_table(\"/path/to/results_files/LCLs_splicing_order_merged_samples_chi_square_vs_distance_results.txt\")\n",
    "order_sig = order_df[(order_df['d']>0.379) & ((order_df['level1']<0.05) | (order_df['level2']<0.05))].reset_index(drop=True)\n",
    "order_genes = order_sig[['gene_name','cell_line']].drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "19f244ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to test for statistical difference in isoform abundance between alleles\n",
    "def compare_isos(gene_name):\n",
    "    \n",
    "    t = 10\n",
    "    results_list = []\n",
    "    \n",
    "    transcripts = esp_df_piv[esp_df_piv['external_gene_name']==gene_name]['transcript_ID'].drop_duplicates().tolist()\n",
    "    \n",
    "    # get all transcript combinations\n",
    "    for combo in itertools.combinations(transcripts, 2):\n",
    "        t1 = combo[0]\n",
    "        t2 = combo[1]\n",
    "    \n",
    "        # retrieve counts for t1 and t2\n",
    "        counts_t1 = esp_df_piv[esp_df_piv['transcript_ID']==t1].reset_index(drop=True)\n",
    "        counts_t2 = esp_df_piv[esp_df_piv['transcript_ID']==t2].reset_index(drop=True)\n",
    "    \n",
    "        # filter for cell lines that have a minimum reads on at least one allele\n",
    "        t1_filt = counts_t1[(counts_t1['maternal']>=t) | (counts_t1['paternal']>=t)].reset_index(drop=True)\n",
    "        t2_filt = counts_t2[(counts_t2['maternal']>=t) | (counts_t2['paternal']>=t)].reset_index(drop=True)\n",
    "    \n",
    "        # Identify cell lines that passed the filter for both transcripts\n",
    "        t1_cell_lines = t1_filt['cell_line'].tolist()\n",
    "        t2_cell_lines = t2_filt['cell_line'].tolist()\n",
    "    \n",
    "        good_cell_lines = [a for a in t1_cell_lines if a in t2_cell_lines]\n",
    "    \n",
    "        # test for statistically significant differences\n",
    "        for ind in good_cell_lines:\n",
    "            t1_p = int(t1_filt[t1_filt['cell_line']==ind]['paternal'])\n",
    "            t2_p = int(t2_filt[t2_filt['cell_line']==ind]['paternal'])\n",
    "            t1_m = int(t1_filt[t1_filt['cell_line']==ind]['maternal'])\n",
    "            t2_m = int(t2_filt[t2_filt['cell_line']==ind]['maternal'])\n",
    "        \n",
    "            OR = stats.fisher_exact([[t1_p,t2_p],[t1_m,t2_m]])[0]\n",
    "            pval = stats.fisher_exact([[t1_p,t2_p],[t1_m,t2_m]])[1]\n",
    "            results_list.append([ind, gene_name, t1, t2, t1_p, t2_p, t1_m, t2_m, OR, pval])\n",
    "    \n",
    "    if len(results_list) > 0:\n",
    "    \n",
    "        results_df = pd.DataFrame(results_list)\n",
    "        results_df.columns = ['cell_line','gene_name','transcript1','transcript2','count1_P','count2_P','count1_M','count2_M','odds_ratio','pvalue']\n",
    "    \n",
    "        return(results_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "786468d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 96/96 [1:29:32<00:00, 55.97s/it]\n"
     ]
    }
   ],
   "source": [
    "# Apply function defined above\n",
    "df_list = []\n",
    "\n",
    "for gene_name in tqdm(order_genes['gene_name'].drop_duplicates().tolist()):\n",
    "#for gene_name in tqdm(gene_list_tmp):\n",
    "    gene_df = compare_isos(gene_name)\n",
    "    if gene_df is not None:\n",
    "        df_list.append(gene_df)\n",
    "    \n",
    "final_df = pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ad4e84fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct for multiple testing\n",
    "final_df['FDR'] = multipletests(final_df['pvalue'], alpha=0.05, method='fdr_bh')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9abc0755",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for significant pvalue and odds ratio\n",
    "final_df_sig = final_df[(final_df['FDR']<0.05) & ((final_df['odds_ratio']>2) | (final_df['odds_ratio']<0.5) | (final_df['odds_ratio']==0.00))].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7cc946a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save transcript pairs and use rMATS-long script classify_isoform_differences.py to determine the coordinates of AS events\n",
    "final_df_sig.to_csv(\"/path/to/results_files/LCL_espresso_pairwise_transcript_comparison_2024-08-27.txt\", sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f0cdbba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get pairwise annotation of alternative splicing differences between isoforms from rmats-long (with only significant pairs)\n",
    "iso_df = pd.read_table(\"/path/to/results_files/LCL_isoform_differences.tsv\")\n",
    "esp_gene_names = esp_df[['transcript_ID','external_gene_name']].drop_duplicates().reset_index(drop=True)\n",
    "iso_df = iso_df.merge(esp_gene_names, left_on=['transcript1'], right_on=['transcript_ID'])\n",
    "\n",
    "# Filter out intron retention events (since we're analyzing chromatin RNA), alternative first/last exons and complex events\n",
    "iso_df_sub = iso_df[~iso_df['event'].isin(['COMPLEX','RI','AFE','ALE'])].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1d176717",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify isoforms that are in the same genes for which splicing order was measured in each cell line\n",
    "order_regions = order_df[['gene_name','gene','analyzed_introns','cell_line','d','level1','level2']].drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Merge the stats results with the annotation of AS events\n",
    "final_df_sig_ann = final_df_sig.merge(iso_df_sub, on=['transcript1','transcript2']).merge(order_regions, on=['cell_line','gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ef86117",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get intron coordinates\n",
    "hg38_intron_df = pd.read_table(\"/path/to/annotation_files/hg38_all_intron_features.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2c6951d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to determine whether the AS event overlaps with the intron group for which splicing order was\n",
    "# computed\n",
    "\n",
    "def AS_vs_order(gene, analyzed_introns, event_type, coord):\n",
    "    \n",
    "    window = 50\n",
    "    \n",
    "    # retrieve coordinates of analyzed introns for splicing order\n",
    "    split_introns = [int(a) for a in analyzed_introns.split(\"_\")]\n",
    "    intron1 = split_introns[0]\n",
    "    intron3 = split_introns[2]\n",
    "    gr_start = int(hg38_intron_df[(hg38_intron_df['gene']==gene) & (hg38_intron_df['intron_pos']==intron1)]['start']) - window\n",
    "    gr_end = int(hg38_intron_df[(hg38_intron_df['gene']==gene) & (hg38_intron_df['intron_pos']==intron3)]['end']) + window\n",
    "    \n",
    "    split_coord = coord.split(\";\")\n",
    "    \n",
    "    flag = \"no_overlap\"\n",
    "    \n",
    "    for event in split_coord:\n",
    "        if flag == \"overlap\":\n",
    "            pass\n",
    "        split_event = event.split(\":\")\n",
    "        event_start = int(split_event[1])\n",
    "        event_end = int(split_event[2])\n",
    "        strand = split_event[3]\n",
    "        \n",
    "        if event_type == \"SE\":\n",
    "            if event_start > gr_start and event_end < gr_end:\n",
    "                flag = \"overlap\"\n",
    "            else:\n",
    "                flag = \"no_overlap\"\n",
    "                \n",
    "        elif event_type == \"A5SS\":\n",
    "            if strand == \"+\":\n",
    "                if event_end > gr_start and event_end < gr_end:\n",
    "                    flag = \"overlap\"\n",
    "                else:\n",
    "                    flag = \"no_overlap\"\n",
    "            elif strand == \"-\":\n",
    "                if event_start > gr_start and event_start < gr_end:\n",
    "                    flag = \"overlap\"\n",
    "                else:\n",
    "                    flag = \"no_overlap\"\n",
    "                    \n",
    "        elif event_type == \"A3SS\":\n",
    "            if strand == \"+\":\n",
    "                if event_start > gr_start and event_start < gr_end:\n",
    "                    flag = \"overlap\"\n",
    "                else:\n",
    "                    flag = \"no_overlap\"\n",
    "            elif strand == \"-\":\n",
    "                if event_end > gr_start and event_end < gr_end:\n",
    "                    flag = \"overlap\"\n",
    "                else:\n",
    "                    flag = \"no_overlap\"\n",
    "                    \n",
    "        else:\n",
    "            flag = \"wrong_event_type\"\n",
    "            \n",
    "    return(flag)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "510f4c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to each row\n",
    "final_df_sig_ann['flag'] = final_df_sig_ann.apply(lambda row: AS_vs_order(row.gene, row.analyzed_introns, row.event, row.coordinates),axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0543c602",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for AS events overlapping intron groups and that are not in HLA genes (those have some alignment artifacts\n",
    "# that lead to \"AS\")\n",
    "AS_in_order_genes = final_df_sig_ann[(final_df_sig_ann['flag']=='overlap') & (~final_df_sig_ann['gene_name'].str.contains(\"HLA\"))].drop_duplicates(subset=['coordinates','cell_line'])\n",
    "len(AS_in_order_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d0773c32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get AS events in genes that showed allele-specific splicing order\n",
    "AS_in_sig_order_genes = AS_in_order_genes[(AS_in_order_genes['d']>0.379) & ((AS_in_order_genes['level1']<0.05) | (AS_in_order_genes['level2']<0.05))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "965cb0d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript1</th>\n",
       "      <th>transcript2</th>\n",
       "      <th>count1_P</th>\n",
       "      <th>count2_P</th>\n",
       "      <th>count1_M</th>\n",
       "      <th>count2_M</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>...</th>\n",
       "      <th>event</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>gene</th>\n",
       "      <th>analyzed_introns</th>\n",
       "      <th>d</th>\n",
       "      <th>level1</th>\n",
       "      <th>level2</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>GM18510</td>\n",
       "      <td>FCER2</td>\n",
       "      <td>ENST00000346664</td>\n",
       "      <td>ESPRESSO:19:14393:5</td>\n",
       "      <td>242</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>10</td>\n",
       "      <td>inf</td>\n",
       "      <td>7.294634e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>A5SS</td>\n",
       "      <td>19:7697527:7697589:-;19:7697401:7697589:-</td>\n",
       "      <td>ENST00000346664</td>\n",
       "      <td>FCER2</td>\n",
       "      <td>NM_001207019.2</td>\n",
       "      <td>4_3_2</td>\n",
       "      <td>0.759306</td>\n",
       "      <td>1.055303e-04</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4324</th>\n",
       "      <td>GM18501</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>ENST00000370751</td>\n",
       "      <td>ESPRESSO:1:1108:25</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>105</td>\n",
       "      <td>33</td>\n",
       "      <td>0.314286</td>\n",
       "      <td>1.003321e-02</td>\n",
       "      <td>...</td>\n",
       "      <td>SE</td>\n",
       "      <td>1:78620944:78620971:+</td>\n",
       "      <td>ENST00000370751</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>NM_006820.3</td>\n",
       "      <td>1_2_3</td>\n",
       "      <td>0.448073</td>\n",
       "      <td>4.496140e-02</td>\n",
       "      <td>7.639776e-01</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4326</th>\n",
       "      <td>GM18501</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>ENST00000370751</td>\n",
       "      <td>ESPRESSO:1:1108:51</td>\n",
       "      <td>14</td>\n",
       "      <td>32</td>\n",
       "      <td>105</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.078529e-22</td>\n",
       "      <td>...</td>\n",
       "      <td>SE</td>\n",
       "      <td>1:78627906:78628393:+</td>\n",
       "      <td>ENST00000370751</td>\n",
       "      <td>IFI44L</td>\n",
       "      <td>NM_006820.3</td>\n",
       "      <td>1_2_3</td>\n",
       "      <td>0.448073</td>\n",
       "      <td>4.496140e-02</td>\n",
       "      <td>7.639776e-01</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4505</th>\n",
       "      <td>GM18501</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>ENST00000531065</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>inf</td>\n",
       "      <td>2.837672e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>A5SS</td>\n",
       "      <td>16:1963653:1964365:-;16:1964276:1964365:-</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>NM_002952.3</td>\n",
       "      <td>3_2_1</td>\n",
       "      <td>1.010153</td>\n",
       "      <td>2.461161e-05</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4513</th>\n",
       "      <td>GM18853</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>ENST00000531065</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.034174e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>A5SS</td>\n",
       "      <td>16:1963653:1964365:-;16:1964276:1964365:-</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>NM_002952.3</td>\n",
       "      <td>3_2_1</td>\n",
       "      <td>1.196642</td>\n",
       "      <td>2.256142e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4521</th>\n",
       "      <td>GM19209</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>ENST00000531065</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>inf</td>\n",
       "      <td>4.087260e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>A5SS</td>\n",
       "      <td>16:1963653:1964365:-;16:1964276:1964365:-</td>\n",
       "      <td>ENST00000527871</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>NM_002952.3</td>\n",
       "      <td>3_2_1</td>\n",
       "      <td>1.141695</td>\n",
       "      <td>9.420491e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>GM19152</td>\n",
       "      <td>SBF1</td>\n",
       "      <td>ESPRESSO:22:20218:32</td>\n",
       "      <td>ESPRESSO:22:20218:33</td>\n",
       "      <td>19</td>\n",
       "      <td>29</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>2.639141e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>SE</td>\n",
       "      <td>22:50457034:50457111:-</td>\n",
       "      <td>ESPRESSO:22:20218:32</td>\n",
       "      <td>SBF1</td>\n",
       "      <td>NM_002972.3</td>\n",
       "      <td>30_29_28</td>\n",
       "      <td>0.974978</td>\n",
       "      <td>3.178884e-01</td>\n",
       "      <td>3.428553e-12</td>\n",
       "      <td>overlap</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     cell_line gene_name           transcript1           transcript2  \\\n",
       "426    GM18510     FCER2       ENST00000346664   ESPRESSO:19:14393:5   \n",
       "4324   GM18501    IFI44L       ENST00000370751    ESPRESSO:1:1108:25   \n",
       "4326   GM18501    IFI44L       ENST00000370751    ESPRESSO:1:1108:51   \n",
       "4505   GM18501      RPS2       ENST00000527871       ENST00000531065   \n",
       "4513   GM18853      RPS2       ENST00000527871       ENST00000531065   \n",
       "4521   GM19209      RPS2       ENST00000527871       ENST00000531065   \n",
       "4531   GM19152      SBF1  ESPRESSO:22:20218:32  ESPRESSO:22:20218:33   \n",
       "\n",
       "      count1_P  count2_P  count1_M  count2_M  odds_ratio        pvalue  ...  \\\n",
       "426        242         0       223        10         inf  7.294634e-04  ...   \n",
       "4324        14        14       105        33    0.314286  1.003321e-02  ...   \n",
       "4326        14        32       105         0    0.000000  4.078529e-22  ...   \n",
       "4505        27         0         1        14         inf  2.837672e-10  ...   \n",
       "4513        22         0         1        15         inf  1.034174e-09  ...   \n",
       "4521        22         0         0        19         inf  4.087260e-12  ...   \n",
       "4531        19        29        19         1    0.034483  2.639141e-05  ...   \n",
       "\n",
       "      event                                coordinates         transcript_ID  \\\n",
       "426    A5SS  19:7697527:7697589:-;19:7697401:7697589:-       ENST00000346664   \n",
       "4324     SE                      1:78620944:78620971:+       ENST00000370751   \n",
       "4326     SE                      1:78627906:78628393:+       ENST00000370751   \n",
       "4505   A5SS  16:1963653:1964365:-;16:1964276:1964365:-       ENST00000527871   \n",
       "4513   A5SS  16:1963653:1964365:-;16:1964276:1964365:-       ENST00000527871   \n",
       "4521   A5SS  16:1963653:1964365:-;16:1964276:1964365:-       ENST00000527871   \n",
       "4531     SE                     22:50457034:50457111:-  ESPRESSO:22:20218:32   \n",
       "\n",
       "     external_gene_name            gene analyzed_introns         d  \\\n",
       "426               FCER2  NM_001207019.2            4_3_2  0.759306   \n",
       "4324             IFI44L     NM_006820.3            1_2_3  0.448073   \n",
       "4326             IFI44L     NM_006820.3            1_2_3  0.448073   \n",
       "4505               RPS2     NM_002952.3            3_2_1  1.010153   \n",
       "4513               RPS2     NM_002952.3            3_2_1  1.196642   \n",
       "4521               RPS2     NM_002952.3            3_2_1  1.141695   \n",
       "4531               SBF1     NM_002972.3         30_29_28  0.974978   \n",
       "\n",
       "            level1        level2     flag  \n",
       "426   1.055303e-04  1.000000e+00  overlap  \n",
       "4324  4.496140e-02  7.639776e-01  overlap  \n",
       "4326  4.496140e-02  7.639776e-01  overlap  \n",
       "4505  2.461161e-05  1.000000e+00  overlap  \n",
       "4513  2.256142e-08  1.000000e+00  overlap  \n",
       "4521  9.420491e-06  1.000000e+00  overlap  \n",
       "4531  3.178884e-01  3.428553e-12  overlap  \n",
       "\n",
       "[7 rows x 21 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AS_in_sig_order_genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
